### PR TITLE
[vpj] Repush will be able to repush from non-chunked version to chunked version

### PR DESCRIPTION
**Pull request by sixpluszero**
## [vpj] Repush will be able to repush from non-chunked version to chunked version

This PR fixes an issue that when repushing from non-chunked to chunked version. Previously the KafkaInputRecordReader  we assume that chunking flag is the same for source version and the dest version, so we rely on new chunking flag to determine how to deserialize key from the source topic and this lead to broken keys in the new versions and partial updates cannot be applied to it.

## How was this PR tested?

New integration test
Will probably add new unit test to meet coverage.

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java [hunk 1]
```diff
@@ -2844,6 +2845,10 @@ private void handleUpdateRequest(
         // to der-se of keys a couple of times.
         updatedKeyBytes = ChunkingUtils.KEY_WITH_CHUNKING_SUFFIX_SERIALIZER.serializeNonChunkedKey(keyBytes);
       }
+      LogManager.getLogger()
```
Comment (gaojieliu) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java]: Remove this?

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 2]
```diff
@@ -2159,6 +2159,7 @@ private void processStartOfPush(
         StoreVersionState newStoreVersionState = new StoreVersionState();
         newStoreVersionState.sorted = startOfPush.sorted;
         newStoreVersionState.chunked = startOfPush.chunked;
+        LOGGER.info("DEBUGGING CHUNKING CONFIG SOP->SVS: " + newStoreVersionState.chunked + " " + versionNumber);
```
Comment (gaojieliu) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java]: Remove this?

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/storage/chunking/SingleGetChunkingAdapter.java [hunk 3]
```diff
@@ -48,6 +50,10 @@ public static ValueRecord get(
     ByteBuffer keyBuffer = isChunked
         ? ChunkingUtils.KEY_WITH_CHUNKING_SUFFIX_SERIALIZER.serializeNonChunkedKeyAsByteBuffer(key)
         : ByteBuffer.wrap(key);
+    LogManager.getLogger()
```
Comment (gaojieliu) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/storage/chunking/SingleGetChunkingAdapter.java]: Remove this?

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VeniceReducer.java [hunk 4]
```diff
@@ -390,6 +390,7 @@ private VeniceWriter<byte[], byte[], byte[]> createBasicVeniceWriter() {
     VeniceWriterFactory veniceWriterFactoryFactory = new VeniceWriterFactory(writerProps);
     boolean chunkingEnabled = props.getBoolean(VeniceWriter.ENABLE_CHUNKING, false);
     boolean rmdChunkingEnabled = props.getBoolean(VeniceWriter.ENABLE_RMD_CHUNKING, false);
+    LOGGER.info("DEBUGGING REDUCER: " + chunkingEnabled + " " + rmdChunkingEnabled + " " + mapReduceJobId.toString());
```
Comment (gaojieliu) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/VeniceReducer.java:393]: Remove this?

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputDictTrainer.java [hunk 5]
```diff
@@ -32,13 +32,15 @@
...
-    private CompressionStrategy sourceVersionCompressionStrategy;
+    private final String kafkaInputBroker;
+    private final String topicName;
+    private final String keySchema;
+    private final Properties sslProperties;
+    private final int compressionDictSize;
+    private final int dictSampleSize;
+    private final CompressionStrategy sourceVersionCompressionStrategy;
+
+    private final boolean sourceVersionChunkingEnabled;
```
Comment (gaojieliu) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputDictTrainer.java:43]: Why is this change required?

Currently, this class has logic to skip the chunked messages by checking the schema id:
https://github.com/linkedin/venice/blob/main/clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputDictTrainer.java#L217

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputRecordReader.java [hunk 6]
```diff
@@ -171,7 +172,9 @@ public boolean next(KafkaInputMapperKey key, KafkaInputMapperValue value) throws
         MessageType messageType = MessageType.valueOf(kafkaMessageEnvelope);
 
         key.offset = pubSubMessage.getOffset();
-        if (isChunkingEnabled) {
+        LOGGER.info(
```
Comment (gaojieliu) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputRecordReader.java]: Remove this?

File: services/venice-server/src/main/java/com/linkedin/venice/listener/StorageReadRequestsHandler.java [hunk 7]
```diff
@@ -410,6 +410,7 @@ private ReadResponse handleSingleGetRequest(GetRouterRequest request) {
 
     AbstractStorageEngine storageEngine = getStorageEngine(topic);
     boolean isChunked = storageEngine.isChunked();
+    LOGGER.info("DEBUGGING GET: IS CHUNKED: " + isChunked);
```
Comment (gaojieliu) [inline services/venice-server/src/main/java/com/linkedin/venice/listener/StorageReadRequestsHandler.java:413]: Remove this?

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputDictTrainer.java (hunk 5 above)
Comment (gaojieliu) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputDictTrainer.java:43]: I guess this is good to keep this change to be consistent, so that KafkaInputRecordReader will be able to execute some chunking logic, and based on the current logic, the outcome is same.

File: internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/PartialUpdateTest.java [hunk 8]
```diff
@@ -140,6 +142,139 @@ public void setUp() {
     this.parentController = parentControllers.get(0);
   }
 
+  @Test
+  public void testReproduceError() throws IOException {
```
Comment (gaojieliu) [inline internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/PartialUpdateTest.java:146]: The test name is confusing.
Is this test mainly to validate the chunking support in repush?
If yes, can we rename it as something like `testRepushWithChunkingFlagChanged`?

File: internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/PartialUpdateTest.java [hunk 9]
```diff
@@ -140,6 +142,139 @@ public void setUp() {
...
+    Schema keySchema = AvroCompatibilityHelper.parse(loadFileAsString("UserKey.avsc"));
+    Schema valueSchema = AvroCompatibilityHelper.parse(loadFileAsString("UserValue.avsc"));
+    Schema writeComputeSchema = WriteComputeSchemaConverter.getInstance().convertFromValueRecordSchema(valueSchema);
+
+    try (ControllerClient parentControllerClient = new ControllerClient(CLUSTER_NAME, parentControllerUrl)) {
+      assertCommand(
+          parentControllerClient.createNewStore(storeName, "test_owner", keySchema.toString(), valueSchema.toString()));
+      UpdateStoreQueryParams updateStoreParams =
+          new UpdateStoreQueryParams().setStorageQuotaInByte(Store.UNLIMITED_STORAGE_QUOTA)
+              // .setPartitionCount(4)
```
Comment (gaojieliu) [inline internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/PartialUpdateTest.java:158]: Can we remove all the commented lines?

**Review by gaojieliu — CHANGES_REQUESTED** [review]
LG overall, left some minor comments.

File: internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/PartialUpdateTest.java (hunk 9 above)
Comment (sixpluszero) [reply internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/PartialUpdateTest.java:158]: Yeah thanks for the review, I need to cleanup those debug logs/commnets and random names before this is ready.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputDictTrainer.java (hunk 5 above)
Comment (sixpluszero) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputDictTrainer.java:43]: I am actually not doing logic change in this class, but I have to populate the source version chunking info into the KafkaInputRecordReader actually. The logic here is a little bit confusing, it seems like we initiate record reader inside tranDict method (which I feel like they belong to separate logic somehow)

Comment (gaojieliu) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputDictTrainer.java:43]: Yes, `dictTrainer` needs to initialize a record read to read the messages from the source topic to build the dict for repush.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputRecordReader.java [hunk 10]
```diff
@@ -122,7 +121,9 @@ public KafkaInputRecordReader(InputSplit split, JobConf job, Reporter reporter)
     this.startingOffset = inputSplit.getStartingOffset();
     this.currentOffset = inputSplit.getStartingOffset() - 1;
     this.endingOffset = inputSplit.getEndingOffset();
-    this.isChunkingEnabled = job.getBoolean(VeniceWriter.ENABLE_CHUNKING, false);
+    // this.isChunkingEnabled = job.getBoolean(VeniceWriter.ENABLE_CHUNKING, false);
```
Comment (gaojieliu) [inline clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputRecordReader.java:124]: Can we remove this?

**Review by gaojieliu — APPROVED** [review]
LGTM overall.

File: clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputRecordReader.java (hunk 10 above)
Comment (sixpluszero) [reply clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputRecordReader.java]: Just removed it.

**Review by gaojieliu — APPROVED** [review]
